server/aesdsocket.c

```diff
@@ -0,0 +1,379 @@
+
+
+/*----------------------------------------------------------------------------
+ int main(int argc, char **argv)
+ *@brief : Main function
+ *
+ *Parameters : argc, argv
+ *
+ *Return : 0 on completion
+ *
+ *----------------------------------------------------------------------------*/
+int main(int argc, char **argv) 
+{
+    openlog(NULL, 0, LOG_USER);
+    syslog(LOG_INFO, "Starting aesdsocket");
+
+    int daemon_set = 0; //Variable used to determine whether the program should run as a daemon.
+
+    //checking if -d argument is given for daemon. If given, setting the daemon flag
+    if (argc > 1 && strcmp(argv[1], "-d") == 0) 
+    {
+        daemon_set = 1;
+    } 
+    else 
+    {
+        syslog(LOG_ERR, "-d is not passed for daemon");
+    }
+
+    //If daemon flag is set, call the daemon function
+    if (daemon_set) 
+    {
+        if (daemon_func() == -1) 
+        {
+            syslog(LOG_ERR, "Failed to create Daemon");
+        } else 
+        {
+            syslog(LOG_DEBUG, "Successfully created Daemon");
+        }
+    }
+
+    //Checking for SIGTERM and SIGINT signals and calling the signal handler function
+    signal(SIGINT, signal_handler);
+    signal(SIGTERM, signal_handler);
+
+    sockfd = setup_socket();  //Creating a socket
+    if (sockfd == -1) {
+        exit_func();
+        return ERROR_SOCKET_CREATE;
+    }
+
+    syslog(LOG_INFO, "Server is listening on port %d", PORT);
+
+    while (1) 
+    {
+        struct sockaddr_in client_addr;
+        socklen_t client_len = sizeof(client_addr);
+        client_fd = accept(sockfd, (struct sockaddr *)&client_addr, &client_len);  //continuously accepting incoming client connections 
+        if (client_fd == -1) {
+            perror("accept");
+            syslog(LOG_ERR, "Error accepting client connection: %m");
+            continue;
+        }
+
+        handle_client_connection(client_fd);
+    }
+
+    exit_func(); //perform cleanup tasks
+    return 0;
+}
+
+
+/*----------------------------------------------------------------------------
+static void signal_handler(int signo) 
+ *@brief : Function to handle the incomming signals
+ *
+ *Parameters : signal number that triggered the signal handler
+ *
+ *Return : none
+ *
+ *----------------------------------------------------------------------------*/
+static void signal_handler(int signo) 
+{
+    if (signo == SIGINT || signo == SIGTERM)           //Checking Received signal is SIGINT or SIGTERM
+    {
+        syslog(LOG_INFO, "Caught signal, exiting");
+        if (unlink(DATA_FILE) == -1)                   // Delete file path
+        { 
+            syslog(LOG_ERR, "Error removing data file: %m");
+        }
+    	if(close(sockfd) == -1)
+	{
+    	     syslog(LOG_ERR, "Error closing data file");
+        }
+       if(close(client_fd) == -1)
+        {
+    	     syslog(LOG_ERR, "Error closing client fd");
+        }
+        closelog();
+        exit(1);
```

Comment:
@vidhyapalaniappan are all of these safe to do from signal handler?  Consider a design which does cleanup outside the signal handler.  This will become more important for assignment 6.